### PR TITLE
Add RandomSampler

### PR DESCRIPTION
**Pull request by arcondello**
Add a performant random sampler. This is intended to replace `dimod.RandomSampler` as part of https://github.com/dwavesystems/dimod/issues/298.

This implements the same API, but also includes a `time_limit` parameter. This is a feature used by the benchmarking team - this sampler is meant to replace their random batch sampler with one maintained as part of Ocean.

I have also attempted to include timing information following the spec in https://github.com/dwavesystems/dwave-samplers/issues/22.

I kind of hate that, because we list the samplers alphabetically, this one gets put first :shrug:.

### CODE REVIEW

**Comment by arcondello** [comment]
This relies on an unreleased dimod feature. Will open a PR shortly.

File: README.rst [hunk 1]
```diff
@@ -24,13 +24,32 @@ or locally on your CPU.
 *dwave-samplers* implements the following classical algorithms for solving
 :term:`binary quadratic model`\ s (BQM):
 
+* Random: a sampler that draws uniform random samples.
```
Comment (kevinchern) [inline README.rst:27]: Missing backtick symbol i.e, 
```
* `Random` : a sampler that draws uniform random samples.
```

Comment (arcondello) [reply README.rst:27]: Those are for the link. In this case I am not linking to anything. Though open to a good place to link to, https://en.wikipedia.org/wiki/Randomization seemed a bit over general :smile:

File: README.rst [hunk 2]
```diff
@@ -24,13 +24,32 @@ or locally on your CPU.
...
   Boltzmann sampling well suited to finding good solutions of large problems.
 * `Steepest Descent`_: a discrete analogue of gradient descent, often used in
   machine learning, that quickly finds a local minimum.
 * `Tabu`_: a heuristic that employs local search with methods to escape local minima.
 * `Tree Decomposition`_: an exact solver for problems with low treewidth.
 
+Random
+======
+
+Random samplers provide a useful baseline performance comparison. The variable
```
Comment (kevinchern) [inline README.rst:38]: Would this be the place to mention which PRNG is used? e.g., Mersenne-Twister?

Comment (arcondello) [reply README.rst:38]: Could do, though the user can actually set that by passing in a [Generator](https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator). By default we use NumPy's default, which is currently [PCG64](https://numpy.org/doc/stable/reference/random/bit_generators/pcg64.html#numpy.random.PCG64). Though I would want to just say that we use the default from NumPy in case NumPy changes.

File: dwave/samplers/random/cyrandom.pyx [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    cdef const char *capsule_name = "BitGenerator"
+    capsule = rng.bit_generator.capsule
+    if not PyCapsule_IsValid(capsule, capsule_name):
+        raise ValueError("Invalid pointer to anon_func_state")
+    bitgen = <numpy.random.bitgen_t *> PyCapsule_GetPointer(capsule, capsule_name)
+
+    cdef double sampling_start_time = realtime_clock()
+
+    cdef double sampling_stop_time
+    if time_limit < 0:
```
Comment (kevinchern) [inline dwave/samplers/random/cyrandom.pyx:118]: Should this be `<=`?

File: dwave/samplers/random/cyrandom.pyx [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    cdef Py_ssize_t i
+    for i in range(cybqm.num_variables()):
+        state.sample.push_back(bitgen.next_uint32(bitgen.state) % 2)
+
+    if is_spin:
+        # go back through and convert to spin
+        for i in range(state.sample.size()):
+            state.sample[i] = 2 * state.sample[i] - 1
+
+    state.energy = cybqm.data().energy(state.sample.begin())
```
Comment (kevinchern) [inline dwave/samplers/random/cyrandom.pyx:88]: Converting spins and computing energies seem more fitting as a postprocessing step.

File: tests/test_random_sampler.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        bqm = dimod.BinaryQuadraticModel({0: 0.0, 1: 0.0, 2: 0.0},
+                                         {(0, 1): -1.0, (1, 2): 1.0, (0, 2): 1.0},
+                                         1.0,
+                                         dimod.SPIN)
+
+        t = time.perf_counter()
+        sampleset = RandomSampler().sample(bqm, num_reads=10, time_limit=.02)
+        runtime = time.perf_counter() - t
+
+        self.assertTrue(.01 < runtime < .04)
```
Comment (kevinchern) [inline tests/test_random_sampler.py:60]: Should also test info['timing']['sampling_time'] is reasonably close to 0.02.

File: tests/test_random_sampler.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+        dimod.testing.assert_sampleset_energies(sampleset, bqm)
+        self.assertEqual(len(sampleset), 10)
```
Comment (kevinchern) [inline tests/test_random_sampler.py:63]: also `info['num_drawn']`

File: dwave/samplers/random/cyrandom.pyx (hunk 4 above)
Comment (arcondello) [reply dwave/samplers/random/cyrandom.pyx:88]: Only in the case where you're keeping every sample. Otherwise you need the energies in order to keep the population to `num_reads`. Because for "real uses" of the solver you need to keep the population limited, IMO it's part of sampling.

File: dwave/samplers/random/cyrandom.pyx [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        for j in range(cybqm.num_variables()):
+            sample_view[i, j] = samples[i].sample[j]
+
+    sampleset = dimod.SampleSet(record, bqm.variables, info=dict(), vartype=bqm.vartype)
+
+    sampleset.info.update(
+        num_drawn=num_drawn,
+        prepreocessing_time=sampling_start_time-preprocessing_start_time,
+        sampling_time=postprocessing_start_time-sampling_start_time,
+        postprocessing_time=realtime_clock()-preprocessing_start_time,
```
[the remark this answers is not in the document]
Comment (arcondello) [reply dwave/samplers/random/cyrandom.pyx:171]: I am open to having it be nested. Though since we're already changing the names, IMO consistency is not the main reason to do it.

File: dwave/samplers/random/sampler.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+                The maximum sampling time in seconds.
+                If given and non-negative, samples are drawn until ``time_limit``.
+                Only the best ``num_reads`` (or fewer) samples are kept.
+
+        Returns:
+            A sample set.
+            Some additional information is provided in the
+            :attr:`~dimod.SampleSet.info` dictionary:
+
+                * **num_drawn**: The total number of samples generated.
```
[the remark this answers is not in the document]
Comment (arcondello) [reply dwave/samplers/random/sampler.py:103]: Specifically when `time_limit` is provided and when it allows for more reads

File: dwave/samplers/random/cyrandom.pyx (hunk 4 above)
Comment (arcondello) [reply dwave/samplers/random/cyrandom.pyx:88]: Four different combos with their uses:
- both unspecified: I just want to use the defaults
- `time_limit` set, `num_reads` unset: I want the best sample I can find randomly in the given time
- `time_limit` unset, `num_reads` set: I want `num_reads` samples drawn uniformly. If I was to specify the time_limit, they would not be uniform.
- both set, I want the best sample(s) found after `time_limit` and I want to specify the number of returned samples and/or the internal memory used.

Obviously these are not inclusive of use cases but should give some idea.

File: README.rst (hunk 1 above)
Comment (JoelPasvolsky) [reply README.rst:27]: It's backtick plus underscore to link to the new Random section in line 35

Comment (arcondello) [reply README.rst:27]: Got it, I misunderstood. Will do.

File: dwave/samplers/random/cyrandom.pyx (hunk 7 above)
[the remark this answers is not in the document]
Comment (arcondello) [reply dwave/samplers/random/cyrandom.pyx:171]: If we do nest it, and if we're already changing the names, IMO it would look nicer to have
```python
info = dict(timing=dict(
    preprocessing=...,
    sampling=...,
    postprocessing=...,
))
```

File: tests/test_random_sampler.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        response = sampler.sample(bqm, num_reads=10)
+
+        dimod.testing.assert_response_energies(response, bqm)
+
+    def test_kwargs(self):
+        bqm = dimod.BinaryQuadraticModel({}, {}, 0.0, dimod.SPIN)
+        with self.assertWarns(dimod.exceptions.SamplerUnknownArgWarning):
+            RandomSampler().sample(bqm, a=5, b=2)
+
+    def test_time_limit(self):
```
Comment (kevinchern) [inline tests/test_random_sampler.py:48]: Maybe add test cases for different settings:
1. default (no `num_reads`, no `time_limit`)
2. only `num_reads`
3. only `time_limit`

File: dwave/samplers/random/cyrandom.pyx (hunk 7 above)
[the remark this answers is not in the document]
Comment (CatherineCMcGeoch) [reply dwave/samplers/random/cyrandom.pyx:171]: IMHO consistency of timing terminology with QPU and other solvers is a priority.  Otherwise the code to plot data from different solvers on the same graph becomes very unwieldy.   This may require some exceptions and compromises because different solvers structures have different structures, but aiming for maximum comparability would be good thing.   This would also be an argument for generic timing categories based on ``once per input'' and ``once per output'' ,  which are natural comparators, rather than solver-specific functions.

[the remark this answers is not in the document]
Comment (kevinchern) [reply dwave/samplers/random/cyrandom.pyx:171]: +1 for nested timings!

File: tests/test_random_sampler.py (hunk 9 above)
Comment (arcondello) [reply tests/test_random_sampler.py:48]: The default case is covered rather extensively by `@dimod.testing.load_sampler_bqm_tests(RandomSampler)`

**Review by kevinchern — COMMENTED** [review]
Maybe add a few more test cases, otherwise it looks good.

File: dwave/samplers/random/cyrandom.pyx [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        samples.push_back(get_sample(cybqm, bitgen, is_spin))
+
+    cdef Py_ssize_t num_drawn = num_reads
+
+    if time_limit >= 0:
+        while realtime_clock() < sampling_stop_time:
+
+            samples.push_back(get_sample(cybqm, bitgen, is_spin))
+            sort(samples.begin(), samples.end(), comp_state)
+            samples.pop_back()
```
Comment (randomir) [inline dwave/samplers/random/cyrandom.pyx:118]: Perhaps use a heap instead? (for total `O(N logN)` complexity)

Alternatively, you can use insertion sort (for `O(N)` per step here, and total `~O(N^2)`) instead of "IntroSort" (for `O(N logN)` per step here, and total of `~O(N^2 logN)`).

File: dwave/samplers/random/cyrandom.pyx (hunk 7 above)
[the remark this answers is not in the document]
Comment (randomir) [reply dwave/samplers/random/cyrandom.pyx:151]: I would also consider typing the duration variables. Perhaps best to use `datetime.timedelta`. I always need to check if some time variable in Ocean is in seconds, ms or us. Hate that.

File: dwave/samplers/random/cyrandom.pyx (hunk 4 above)
Comment (randomir) [reply dwave/samplers/random/cyrandom.pyx:70]: I would forbid user setting both `num_reads` and `time_limit`, since `time_limit` takes precedence and might return less than `num_reads`.

File: dwave/samplers/random/sampler.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            num_reads: The number of samples to be returned.
+
+            seed:
+                Seed for the random number generator.
+                Passed to :func:`numpy.random.default_rng()`.
+
+            time_limit:
+                The maximum sampling time in seconds.
+                If given and non-negative, samples are drawn until ``time_limit``.
+                Only the best ``num_reads`` (or fewer) samples are kept.
```
Comment (randomir) [inline dwave/samplers/random/sampler.py:96]: Until we figure out how best to extend/generalize `Initialized` sampler interface to handle other termination criteria, I would forbid setting both `num_reads` and `time_limit` here. It's confusing because `time_limit` takes precedence and the sampler might return less then `num_reads`.

A case of setting `time_limit`, but not `num_reads` is also confusing -- we return only one sample after `num_drawn` tries.

Perhaps we should introduce `max_answers` (or `max_samples`) that can be used in combination with `time_limit`. See https://github.com/dwavesystems/dwave-greedy/issues/24 for discussion.

So:
- `num_reads` set: exactly `num_reads` returned, `time_limit` not allowed. `max_answers` ignored.
- `time_limit` set: `max_answers` (which can default to 1 or `inf`/None) is used

With this scheme, I would set `num_reads` by default to `None`, with interpretation on how much that actually is depending on `time_limit`/`max_answers`. Similarly to how we resolve it in the `Initialized` arg parser.

File: releasenotes/notes/random-sampler-a5bacc2b7761222f.yaml [hunk 12]
```diff
@@ -0,0 +1,3 @@
+---
+features:
+  - Add ``RandomSampler`` which generates samples with variable assignments chosen by coin flip.
```
Comment (randomir) [inline releasenotes/notes/random-sampler-a5bacc2b7761222f.yaml:3]: Unless we have a coin flipping device attached to user's machine, I would rephrase this :smile:

**Review by randomir — COMMENTED** [review]
Looks good overall, my only concern is about sampling termination criteria interface.

File: dwave/samplers/random/sampler.py (hunk 11 above)
Comment (arcondello) [reply dwave/samplers/random/sampler.py:96]: I had already changed it so `num_reads` has precedence. And I disagree that they should be mutually exclusive. The solver will now always return `num_reads`. `time_limit` just determines whether it keeps going.

File: dwave/samplers/random/cyrandom.pyx (hunk 4 above)
Comment (arcondello) [reply dwave/samplers/random/cyrandom.pyx:70]: https://github.com/dwavesystems/dwave-samplers/pull/24/commits/947320b1a28522c278a4520ad4182c17ed7f05ab changed the behavior to always return `num_reads`. `time_limit` now just allows the user to keep sampling.

Comment (randomir) [reply dwave/samplers/random/cyrandom.pyx:70]: I was looking at your comment above, and missed that update, thanks.

File: dwave/samplers/random/cyrandom.pyx (hunk 7 above)
[the remark this answers is not in the document]
Comment (randomir) [reply dwave/samplers/random/cyrandom.pyx:151]: QPU timing info could be considered encoded for transport -- and we can decode it as `timedelta` client-side.

File: dwave/samplers/random/cyrandom.pyx (hunk 10 above)
Comment (randomir) [reply dwave/samplers/random/cyrandom.pyx:118]: Or constant factor too big to notice benefits on smaller scale.

Comment (arcondello) [reply dwave/samplers/random/cyrandom.pyx]: Ok, after the change introduced in https://github.com/dwavesystems/dwave-samplers/pull/24/commits/176c25c2e8d86ecf3b51161482248c35644b570f, I was able to get a performance boost at very large `num_reads`. So I will indeed make the change. Thanks, and good suggestion!

Comment (arcondello) [reply dwave/samplers/random/cyrandom.pyx]: For posterity I used
```python
bqm = dimod.generators.gnp_random_bqm(1000, .5, 'BINARY')
num_reads=10000
```

File: dwave/samplers/random/sampler.py (hunk 11 above)
Comment (randomir) [reply dwave/samplers/random/sampler.py:96]: Ok, so it looks an implied AND between `==num_reads` and `>=time_limit`:
- sampler will always return `num_reads`,
- sampler will run for at least `time_limit`?

I find this awkward and inconsistent with `Initialized` (`num_reads` adapts to `initial_states` if states are user-specified and reads are not) and termination criteria in dwave-hybrid/`Kerberos` (time/convergence/energy/iteration count are OR-ed - we stop as soon as one condition is met).

Comment (arcondello) [reply dwave/samplers/random/sampler.py:96]: I don't think it needs to be consistent with `Initialized` because it doesn't have any concept of an initial state.

My understanding from benchmarking is that they prefer to give the sampler a fixed runtime, so we need some sort of time-based exit criteria.

So, we can either do treat `time_limit` as the most important, and then have a `max_answers`/`max_samples` as the secondary criteria. Or we can treat `num_reads` and the most important and treat `time_limit` as the secondary criteria when provided.

I much prefer the latter for a few reasons:
1) I don't know how to choose a good default `time_limit`, since the amount of solving will vary a lot for different BQMs.
2) for even moderately long time limits, the `num_reads` will be a tiny fraction. So while it is possible for `time_limit` to be exceeded for large `num_reads`, most typical uses will not run into that. It's also addressed by making `num_reads` default 1. 
3) `num_reads` is consistent with most of our other samples.

Comment (randomir) [reply dwave/samplers/random/sampler.py:96]: I find disjunction (OR) between termination criteria much simpler and cleaner.

- If user sets `num_reads` (and not `time_limit`), they obviously don't care about run time, and just want to get specific number of samples back.

- If user sets `time_limit` (and not `num_reads`), they care about run time, although it's not clear what's the number of samples they want. But if they haven't specified the number, it makes sense to return _all_ samples generated during runtime. Here's where `max_samples`/`max_answers` comes into play.

- In this framework, setting both `num_reads` and `time_limit` would run until one of them is satisfied.

Comment (arcondello) [reply dwave/samplers/random/sampler.py:96]: Ok, how about I just make two samplers. One that take a `time_limit` and `max_answers` and one that takes `num_reads`. If we're going to make the arguments disjoint, we might as well make that clear at the sampler level.

Personally I think that's way overkill, but it seems like no one else agrees with me :smile:

Comment (arcondello) [reply dwave/samplers/random/sampler.py:96]: > If user sets time_limit (and not num_reads), they care about run time, although it's not clear what's the number of samples they want. But if they haven't specified the number, it makes sense to return all samples generated during runtime. Here's where max_samples/max_answers comes into play.

The problem is that we generate a lot of samples. Way more than can easily fit in memory. So returning _all_ is not practical for many problems.

> In this framework, setting both num_reads and time_limit would run until one of them is satisfied.

This is what I had before.

Comment (kevinchern) [reply dwave/samplers/random/sampler.py:96]: > So, we can either do treat time_limit as the most important, and then have a max_answers/max_samples as the secondary criteria. Or we can treat num_reads and the most important and treat time_limit as the secondary criteria when provided.

If we're deciding between the two, I also favour the latter for similar reasons. I do think Radomir's proposal is more intuitive and gives more control or the behaviour of the sampler.

Another point to consider is if we add a similar feature for Neal, a consistent behaviour across classes of samplers would be desirable. And in the case of Neal, I think it's reasonable to set a time limit retrieve all answers it came across.

Comment (randomir) [reply dwave/samplers/random/sampler.py:96]: Extending sampler interface with `time_limit` (and `max_answers` or whatever we call it) would be a good candidate for a follow-up PR, just to keep concerns separated. But since we had all this discussion here, we might as well do it here and then generalize in a follow-up.

**Comment by arcondello** [comment]
Ok, refactored. Now both `num_reads` and `time_limit` are termination criteria. Added `max_num_samples` to limit the population of samples kept internally.

I also have dropped the timing info, pending a decision on #22. We can restore it in a followup PR.

File: README.rst [hunk 13]
```diff
@@ -24,13 +24,32 @@ or locally on your CPU.
...
+>>> sampler = RandomSampler()
+
+Create a random binary quadratic model.
+
+>>> import dimod
+>>> bqm = dimod.generators.gnp_random_bqm(100, .5, 'BINARY')
+
+Get the best 5 sample found in .1 seconds
+
+>>> sampleset = sampler.sample(bqm, time_limit=.1, max_num_samples=5)
```
Comment (randomir) [inline README.rst:51]: Perhaps include ~`.info['num_drawn']`~ `.info['num_reads']` here to clarify many more were actually generated.

File: dwave/samplers/random/sampler.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+        Args:
+            bqm: Binary quadratic model to be sampled from.
+
+            num_reads:
+                The maximum number of random samples to be drawn.
+                If neither ``num_reads`` nor ``time_limit`` are provided,
+                ``num_reads`` is set to 1.
+                If ``time_limit`` is provided, there is no limit imposed on
+                the number of reads.
```
Comment (randomir) [inline dwave/samplers/random/sampler.py:100]: Isn't the limit set by `max_num_samples` then?

Comment (arcondello) [reply dwave/samplers/random/sampler.py:100]: That's the limit on the population, it will take unlimited reads up to the time limit.
